### PR TITLE
updates to grammer and spelling

### PR DESCRIPTION
**Pull request by pschou**
Updates to the basics.md for Prometheus usage documentation.

### CODE REVIEW

File: docs/querying/basics.md [hunk 1]
```diff
@@ -165,12 +194,12 @@ following units:
 * `s` - seconds
 * `m` - minutes
 * `h` - hours
-* `d` - days - assuming a day has always 24h
```
Comment (roidelapluie) [inline docs/querying/basics.md:168]: This is needed. It means that we do not take DST, Leap hours, into consideration.

File: docs/querying/basics.md [hunk 2]
```diff
@@ -165,12 +194,12 @@ following units:
...
-* `w` - weeks - assuming a week has always 7d
-* `y` - years - assuming a year has always 365d
+* `d` - days - assuming a day has 24h
+* `w` - weeks - assuming a week has 7d
+* `y` - years - assuming a year has 365d
 
-Time durations can be combined, by concatenation. Units must be ordered from the
-longest to the shortest. A given unit must only appear once in a time duration.
+Time durations can be combined by concatenation. Units must be ordered from the
+longest to the shortest. A given unit must only appear once in the time duration.
```
Comment (roidelapluie) [inline docs/querying/basics.md:202]: Why changing this?

File: docs/querying/basics.md [hunk 3]
```diff
@@ -204,6 +233,10 @@ The same works for range vectors. This returns the 5-minute rate that
 
     rate(http_requests_total[5m] offset 1w)
 
+When querying for samples in the past, a negative offset will enable temporal comparisons forward in time:
```
Comment (roidelapluie) [inline docs/querying/basics.md:236]: This is not part of this PR.

File: docs/querying/basics.md [hunk 4]
```diff
@@ -218,30 +251,31 @@ For example, the following expression returns the value of
 Note that the `@` modifier always needs to follow the selector
 immediately, i.e. the following would be correct:
 
-    sum(http_requests_total{method="GET"} @ 1609746000) // GOOD.
+    sum( http_requests_total{method="GET"} @ 1609746000 ) // GOOD.
```
Comment (roidelapluie) [inline docs/querying/basics.md:254]: Please do not change the formatting here, I think it reads better without the spaces.

File: docs/querying/basics.md [hunk 5]
```diff
@@ -218,30 +251,31 @@ For example, the following expression returns the value of
...
 
 While the following would be *incorrect*:
 
-    sum(http_requests_total{method="GET"}) @ 1609746000 // INVALID.
+    sum( http_requests_total{method="GET"} ) @ 1609746000 // INVALID.
```
Comment (roidelapluie) [inline docs/querying/basics.md:258]: Please do not change the formatting here, I think it reads better without the spaces.

File: docs/querying/basics.md [hunk 6]
```diff
@@ -218,30 +251,31 @@ For example, the following expression returns the value of
...
 `http_requests_total` had at `2021-01-04T07:40:00+00:00`:
 
     rate(http_requests_total[5m] @ 1609746000)
 
-`@` modifier supports all representation of float literals described
-above within the limits of `int64`. It can also be used along
-with `offset` modifier where the offset is applied relative to the `@`
-modifier time irrespective of which modifier is written first.
-These 2 queries will produce the same result.
+`@` modifier supports all representation of numeric literals described
```
Comment (roidelapluie) [inline docs/querying/basics.md:265]: nan and +inf are float too, but nan is not numeric.

File: docs/querying/basics.md [hunk 7]
```diff
@@ -218,30 +251,31 @@ For example, the following expression returns the value of
...
+These two queries will produce the same result.
 
     # offset after @
     http_requests_total @ 1609746000 offset 5m
     # offset before @
     http_requests_total offset 5m @ 1609746000
 
 This modifier is disabled by default since it breaks the invariant that PromQL
-does not look ahead of the evaluation time for samples. It can be enabled by setting
+does not look ahead of the evaluation time for samples. This can be enabled by setting
```
Comment (roidelapluie) [inline docs/querying/basics.md:278]: Why repeating this?

**Review by roidelapluie — COMMENTED** [review]
I have added some comments but I feel like @RichiH or @oddlittlebird 's input would be great here for the wording part.

File: docs/querying/basics.md (hunk 1 above)
Comment (pschou) [reply docs/querying/basics.md:168]: I agree, but the adverb "always" should be in front of the verb it is modifying, in this case: 'has'.

assuming a day always has 24h

File: docs/querying/basics.md (hunk 2 above)
Comment (pschou) [reply docs/querying/basics.md:202]: These are small, but here are my thoughts:

It's an independent clause (a self-sustaining thought), which is then followed by a modifier.

The word "the" instead of "a" overloads subject defined in the first sentence.

File: docs/querying/basics.md (hunk 3 above)
Comment (pschou) [reply docs/querying/basics.md:236]: Yup, this is my error.  I did a cp and didn't remove this update.

File: docs/querying/basics.md (hunk 4 above)
Comment (pschou) [reply docs/querying/basics.md:254]: I would really like to pull attention out to the new placement of the parenthesis.  I agree that the spaces distract.

What do you think about this:  Could we color the outer sum parenthesis to draw attention to them?

File: docs/querying/basics.md (hunk 5 above)
Comment (pschou) [reply docs/querying/basics.md:258]: Same as above, agreed.

File: docs/querying/basics.md (hunk 7 above)
Comment (pschou) [reply docs/querying/basics.md:278]: Thought:

Replacing the word "This" in place of "It" continues the use of the established subject.

Comment (pschou) [reply docs/querying/basics.md:278]: Yes, you are right, "it" is better as it is a continuation of the previous subject.

**Comment by roidelapluie** [comment]
cc @RichiH @hdost

**Review by hdost — COMMENTED** [review]
Looks good from my perspective

**Comment by beorn7** [comment]
Hello from the bug scrub. @roidelapluie wanted to massage thin into merging, but obviously he forgot. @krajorama will take this now and "fix the fixes".

File: docs/querying/basics.md [hunk 8]
```diff
@@ -281,33 +316,33 @@ PromQL supports line comments that start with `#`. Example:
...
+subsequently ingested for that time series, they will be returned as expected.
 
-If no sample is found (by default) 5 minutes before a sampling timestamp,
-no value is returned for that time series at this point in time. This
-effectively means that time series "disappear" from graphs at times where their
-latest collected sample is older than 5 minutes or after they are marked stale.
+A time series will go stale when they're no longer exposed / the target no
+longer exists. This effectively means that time series "disappears" from graphs 
+at the times of their latest collected sample, and they will not show up in queries
+over a time interval they are marked stale.
```
Comment (bboreham) [inline docs/querying/basics.md:338]: Both "target no longer exists" and "5 minutes" behaviours are still in Prometheus, although the latter is seen mostly with targets that set explicit timestamps.
But we should document both, not erase one.

File: docs/querying/basics.md [hunk 9]
```diff
@@ -14,7 +14,7 @@ systems via the [HTTP API](api.md).
 
 ## Examples
 
-This document is meant as a reference. For learning, it might be easier to
+This document is a Promethues basic language reference. For learning, it may be easier to
```
Comment (krajorama) [inline docs/querying/basics.md:17]:
```suggestion
This document is a Prometheus basic language reference. For learning, it may be easier to
```

File: docs/querying/basics.md [hunk 10]
```diff
@@ -28,24 +28,23 @@ evaluate to one of four types:
 * **String** - a simple string value; currently unused
```
Comment (krajorama) [inline docs/querying/basics.md:28]: note to self: start another PR due to native histograms experimental feature where there's a new type

File: docs/querying/basics.md [hunk 11]
```diff
@@ -299,33 +331,33 @@ PromQL supports line comments that start with `#`. Example:
 
 ### Staleness
 
-When queries are run, timestamps at which to sample data are selected
-independently of the actual present time series data. This is mainly to support
+The timestamps at which to sample data, during a query, are selected
+independently of the actual present time series data. This method principally supports
 cases like aggregation (`sum`, `avg`, and so on), where multiple aggregated
-time series do not exactly align in time. Because of their independence,
+time series do not precisely precisely in time. Because of their independence,
```
Comment (krajorama) [inline docs/querying/basics.md:337]:
```suggestion
time series do not precisely align in time. Because of their independence,
```

File: docs/querying/basics.md [hunk 12]
```diff
@@ -299,33 +331,33 @@ PromQL supports line comments that start with `#`. Example:
...
+subsequently ingested for that time series, they will be returned as expected.
 
-If no sample is found (by default) 5 minutes before a sampling timestamp,
-no value is returned for that time series at this point in time. This
-effectively means that time series "disappear" from graphs at times where their
-latest collected sample is older than 5 minutes or after they are marked stale.
+A time series will go stale when they're no longer exposed / the target no
+longer exists. This effectively means that time series "disappears" from graphs 
+at the times of their latest collected sample, and they will not show up in queries
+over a time interval they are marked stale.
```
Comment (krajorama) [inline docs/querying/basics.md:353]: Let's restore this

```suggestion
If no sample is found (by default) 5 minutes before a sampling timestamp,
no value is returned for that time series at this point in time. This
effectively means that time series "disappear" from graphs at times where their
latest collected sample is older than 5 minutes or after they are marked stale.
```

File: docs/querying/basics.md (hunk 8 above)
Comment (krajorama) [reply docs/querying/basics.md]: i've removed the change

Comment (krajorama) [reply docs/querying/basics.md]: updated with your suggestion please take a look

**Comment by bboreham** [comment]
Last commit needs signed-off; I can do it since I wrote the text.

**Comment by krajorama** [comment]
> Last commit needs signed-off; I can do it since I wrote the text.

There's more older commits from the original author that are not signed. I think simplest would be for me to just squash open a new PR and make sure to have Author/Co-Author correct.

File: docs/querying/basics.md [hunk 13]
```diff
@@ -46,16 +46,15 @@ _Notes about the experimental native histograms:_
 
 ### String literals
 
-Strings may be specified as literals in single quotes, double quotes or
-backticks.
+Strings literals are designated by single quotes, double quotes or backticks.
```
Comment (bboreham) [inline docs/querying/basics.md:49]:
```suggestion
String literals are designated by single quotes, double quotes or backticks.
```

File: docs/querying/basics.md [hunk 14]
```diff
@@ -87,17 +86,17 @@ Examples:
 
 ### Instant vector selectors
 
-Instant vector selectors allow the selection of a set of time series and a
-single sample value for each at a given timestamp (instant): in the simplest
-form, only a metric name is specified. This results in an instant vector
+Instant vector selectors allow the selection of a set of time series (in a time range query) 
+and a single sample value for each at a given timestamp (in an instant query).  In the simplest
+form, only a metric name is specified, which results in an instant vector
```
Comment (bboreham) [inline docs/querying/basics.md:91]: This change doesn't seem right to me.  The original is trying to define "instant vector" as a thing independent of the query it appears in.  "timestamp (instant)" means the word "timestamp" means the same as the word "instant".

File: docs/querying/basics.md [hunk 15]
```diff
@@ -106,7 +105,7 @@ metric name that also have the `job` label set to `prometheus` and their
 
     http_requests_total{job="prometheus",group="canary"}
 
-It is also possible to negatively match a label value, or to match label values
+It is also possible to negatively match a label value or match label values
```
Comment (bboreham) [inline docs/querying/basics.md:108]: very low-value change

File: docs/querying/basics.md [hunk 16]
```diff
@@ -124,6 +123,38 @@ For example, this selects all `http_requests_total` time series for `staging`,
 Label matchers that match empty label values also select all time series that
 do not have the specific label set at all. It is possible to have multiple matchers for the same label name.
 
+An example dataset may include:
```
Comment (bboreham) [inline docs/querying/basics.md:126]:
```suggestion
For example, given the dataset:
```

File: docs/querying/basics.md [hunk 17]
```diff
@@ -124,6 +123,38 @@ For example, this selects all `http_requests_total` time series for `staging`,
...
+    http_requests_total{replica="rep-b"}
+
+and would exclude:
+
+    http_requests_total{environment="development"}
+
+Regex-matches are fully anchored, meaning the pattern will only match if the whole string matches.
+
+    http_requests_total{replica=~"^rep.*"} # Good!
+    http_requests_total{replica=~"^rep"}   # No match
```
Comment (bboreham) [inline docs/querying/basics.md:146]: This point has already been added on line 116.  Suggest to drop here.

File: docs/querying/basics.md [hunk 18]
```diff
@@ -176,13 +207,15 @@ following units:
 
 * `ms` - milliseconds
 * `s` - seconds
-* `m` - minutes
+* `m` - minutes - assuming a minute always has 60s<sup>1</sup>
```
Comment (bboreham) [inline docs/querying/basics.md:210]: unnecessary

File: docs/querying/basics.md [hunk 19]
```diff
@@ -299,33 +331,34 @@ PromQL supports line comments that start with `#`. Example:
 
 ### Staleness
 
-When queries are run, timestamps at which to sample data are selected
-independently of the actual present time series data. This is mainly to support
+The timestamps at which to sample data, during a query, are selected
+independently of the actual present time series data. This method principally supports
```
Comment (bboreham) [inline docs/querying/basics.md:335]: This statement about timestamps is not universally true, and so we should not make it here.
I think it's true for all range queries, and for instant queries that do not aggregate.

Separately, changing "mainly" to "principally" is unnecessary.

File: docs/querying/basics.md [hunk 20]
```diff
@@ -334,10 +367,10 @@ or aggregated your data sufficiently, switch to graph mode. If the expression
 still takes too long to graph ad-hoc, pre-record it via a [recording
 rule](../configuration/recording_rules.md#recording-rules).
 
-This is especially relevant for Prometheus's query language, where a bare
+This consideration is especially relevant for Prometheus's query language, where a bare
```
Comment (bboreham) [inline docs/querying/basics.md:370]: unnecessary

File: docs/querying/basics.md [hunk 21]
```diff
@@ -334,10 +367,10 @@ or aggregated your data sufficiently, switch to graph mode. If the expression
...
 metric name selector like `api_http_requests_total` could expand to thousands
-of time series with different labels. Also keep in mind that expressions which
-aggregate over many time series will generate load on the server even if the
-output is only a small number of time series. This is similar to how it would
+of time series with different labels. Also, keep in mind that expressions that
+aggregate over many number of time series samples will generate load on the server even if the
```
Comment (bboreham) [inline docs/querying/basics.md:373]: "many number" is not English.

File: docs/querying/basics.md [hunk 22]
```diff
@@ -334,10 +367,10 @@ or aggregated your data sufficiently, switch to graph mode. If the expression
...
+output is only a small number of time series samples. This is similar to how it would
```
Comment (bboreham) [inline docs/querying/basics.md:374]: incorrect change

**Review by bboreham — COMMENTED** [review]
I didn't read this closely before.  I have many issues with the changes proposed, particularly where a longer word is substituted for a shorter one.

File: docs/querying/basics.md (hunk 17 above)
Comment (krajorama) [inline docs/querying/basics.md:146]:
```suggestion
```

File: docs/querying/basics.md (hunk 20 above)
Comment (krajorama) [inline docs/querying/basics.md:370]:
```suggestion
This is especially relevant for Prometheus's query language, where a bare
```

File: docs/querying/basics.md (hunk 22 above)
Comment (krajorama) [inline docs/querying/basics.md:374]:
```suggestion
of time series with different labels. Also, keep in mind that expressions that
aggregate over many time series will generate load on the server even if the
output is only a small number of time series. This is similar to how it would
```

File: docs/querying/basics.md [hunk 23]
```diff
@@ -46,16 +46,15 @@ _Notes about the experimental native histograms:_
...
-backticks.
+String literals are designated by single quotes, double quotes or backticks.
 
 PromQL follows the same [escaping rules as
-Go](https://golang.org/ref/spec#String_literals). In single or double quotes a
+Go](https://golang.org/ref/spec#String_literals). For string literals in single or double quotes, a
 backslash begins an escape sequence, which may be followed by `a`, `b`, `f`,
-`n`, `r`, `t`, `v` or `\`. Specific characters can be provided using octal
-(`\nnn`) or hexadecimal (`\xnn`, `\unnnn` and `\Unnnnnnnn`).
+`n`, `r`, `t`, `v` or `\`.  Characters can be produced using octal
```
Comment (krajorama) [inline docs/querying/basics.md:54]:
```suggestion
`n`, `r`, `t`, `v` or `\`.  Specific characters can be provided using octal
```

File: docs/querying/basics.md (hunk 14 above)
Comment (krajorama) [reply docs/querying/basics.md:91]: I guess for some the word instant is overloaded, so we could say `(instant in time)` or just not have it there?

File: docs/querying/basics.md [hunk 24]
```diff
@@ -124,6 +123,33 @@ For example, this selects all `http_requests_total` time series for `staging`,
...
+
+    http_requests_total{environment="development"}
+
+Multiple matchers can be used for the same label name; they all must pass for a result to be returned. 
+
+The query:
+
+    http_requests_total{replica!="rep-a",replica=~"rep.*"}
+
+Would them match:
```
Comment (krajorama) [inline docs/querying/basics.md:149]:
```suggestion
Would then match:
```

File: docs/querying/basics.md [hunk 25]
```diff
@@ -87,17 +86,17 @@ Examples:
 
 ### Instant vector selectors
 
-Instant vector selectors allow the selection of a set of time series and a
-single sample value for each at a given timestamp (instant): in the simplest
-form, only a metric name is specified. This results in an instant vector
+Instant vector selectors allow the selection of a set of time series (in a time range query) 
+and a single sample value for each at a given timestamp (in an instant query).  In the simplest
```
Comment (krajorama) [inline docs/querying/basics.md:90]:
```suggestion
Instant vector selectors allow the selection of a set of time series and a
single sample value for each at a given timestamp (instant in time).  In the simplest
```

File: docs/querying/basics.md (hunk 18 above)
Comment (krajorama) [reply docs/querying/basics.md:205]: isn't the point that there is no account for leap seconds ?

File: docs/querying/basics.md [hunk 26]
```diff
@@ -299,33 +326,34 @@ PromQL supports line comments that start with `#`. Example:
 
 ### Staleness
 
-When queries are run, timestamps at which to sample data are selected
-independently of the actual present time series data. This is mainly to support
+The timestamps at which to sample data, during a query, are selected
+independently of the actual present time series data. This method principally supports
```
Comment (krajorama) [inline docs/querying/basics.md:330]:
```suggestion
independently of the actual present time series data. This is mainly to support
```

File: docs/querying/basics.md (hunk 14 above)
Comment (krajorama) [reply docs/querying/basics.md:91]: Or point in time, which might be more English

File: docs/querying/basics.md (hunk 19 above)
Comment (krajorama) [reply docs/querying/basics.md]: Reading through the rest of the document and after watching https://www.youtube.com/watch?v=xIAEEQwUBXQ&t=497s the nth time I think we should just call out the fact that instant vector selectors and range vector selectors work differently and not try to explain here again, because that is redundant and can lead to faulty documentation later if things get out of sync, making proposal.

File: docs/querying/basics.md [hunk 27]
```diff
@@ -299,33 +326,34 @@ PromQL supports line comments that start with `#`. Example:
...
 
-When queries are run, timestamps at which to sample data are selected
+The timestamps at which to sample data, during a query, are selected
 independently of the actual present time series data. This is mainly to support
 cases like aggregation (`sum`, `avg`, and so on), where multiple aggregated
-time series do not exactly align in time. Because of their independence,
+time series do not precisely align in time. Because of their independence,
 Prometheus needs to assign a value at those timestamps for each relevant time
-series. It does so by simply taking the newest sample before this timestamp.
+series. It does so by taking the newest sample before this timestamp.
```
Comment (krajorama) [inline docs/querying/basics.md:334]:
```suggestion
series. It does so by taking the newest sample before this timestamp in the lookback period.
The lookback period is 5 minutes by default.
```

File: docs/querying/basics.md [hunk 28]
```diff
@@ -46,16 +46,15 @@ _Notes about the experimental native histograms:_
...
+String literals are designated by single quotes, double quotes or backticks.
 
 PromQL follows the same [escaping rules as
-Go](https://golang.org/ref/spec#String_literals). In single or double quotes a
+Go](https://golang.org/ref/spec#String_literals). For string literals in single or double quotes, a
 backslash begins an escape sequence, which may be followed by `a`, `b`, `f`,
-`n`, `r`, `t`, `v` or `\`. Specific characters can be provided using octal
-(`\nnn`) or hexadecimal (`\xnn`, `\unnnn` and `\Unnnnnnnn`).
+`n`, `r`, `t`, `v` or `\`.  Specific characters can be provided using octal
+(`\nnn`) or hexadecimal (`\xnn`, `\unnnn` and `\Unnnnnnnn`) notations.
```
Comment (bboreham) [inline docs/querying/basics.md:55]:
```suggestion
(`\nnn`) or hexadecimal (`\xnn`, `\unnnn` and `\Unnnnnnnn`) notations.
```

File: docs/querying/basics.md (hunk 18 above)
Comment (bboreham) [reply docs/querying/basics.md:205]: Yes, who knows about leap seconds?  There haven't been any since 2016; I understand there will never be any more leap seconds. [Wikipedia](https://en.wikipedia.org/wiki/Leap_second#International_proposals_for_elimination_of_leap_seconds)

We could perhaps put on some 'concepts' page the detailed definition of time that Prometheus is based on; I do not think it belongs on the "querying basics" page.

File: docs/querying/basics.md [hunk 29]
```diff
@@ -299,33 +326,34 @@ PromQL supports line comments that start with `#`. Example:
 
 ### Staleness
 
-When queries are run, timestamps at which to sample data are selected
+The timestamps at which to sample data, during a query, are selected
 independently of the actual present time series data. This is mainly to support
```
[the remark this answers is not in the document]
Comment (bboreham) [reply docs/querying/basics.md:330]: I don't think this is correct.  I think an instant query will give you actual timestamps from the underlying series, unless you aggregate the data in which case I don't know what it will do.
A range query never gives timestamps from the underlying series; they are always a multiple of `step` away from `end`.
(and subqueries are a multiple of their `step` away from the epoch, although I don't understand why)

File: docs/querying/basics.md [hunk 30]
```diff
@@ -176,13 +202,15 @@ following units:
 
 * `ms` - milliseconds
 * `s` - seconds
-* `m` - minutes
+* `m` - minutes - assuming a minute always has 60s<sup>1</sup>
```
Comment (krajorama) [inline docs/querying/basics.md:205]:
```suggestion
* `m` - minutes
```

File: docs/querying/basics.md (hunk 29 above)
[the remark this answers is not in the document]
Comment (krajorama) [reply docs/querying/basics.md:330]: Arrrrrggg, I linked the right thing but wrote `query selector` instead of `vector selector` . Updated suggestion. Perhaps it would help if these weren't so similar in the first place.

File: docs/querying/basics.md [hunk 31]
```diff
@@ -249,11 +280,11 @@ The same works for range vectors. This returns the 5-minute rate that
 
     rate(http_requests_total[5m] @ 1609746000)
 
-The `@` modifier supports all representation of float literals described
-above within the limits of `int64`. It can also be used along
-with the `offset` modifier where the offset is applied relative to the `@`
-modifier time irrespective of which modifier is written first.
-These 2 queries will produce the same result.
+The `@` modifier supports all representations of numeric literals described above.
+It works with the `offset` modifier where the offset is applied relative to the `@`
```
Comment (bboreham) [inline docs/querying/basics.md:284]: "works" implies it is typical to combine them, whereas "can be used" says it is possible but not the norm.
In this case I think "can be used" is better.

Also I think a new paragraph at "It" would help to separate the description of `@` from the combination of `@` and `offset` which follows.

**Review by bboreham — APPROVED** [review]
Approving, this seems all fine to me.  One nit but I don't really mind.
